contracts/DelegatedMigrator.sol

```diff
@@ -0,0 +1,134 @@
+pragma solidity ^0.5.16;
+import "./Owned.sol";
+import "./AddressResolver.sol";
+import "./Proxy.sol";
+import "./SystemStatus.sol";
+
+interface IMigration {
+    // pass the index to handle situations where the migration takes a few transactions to process
+    function migrate(uint256 index) external;
+
+    function numOfScripts() external view returns (uint);
+}
+
+contract DelegatedMigrator is Owned {
+    uint waitingPeriod;
+
+    struct Migration {
+        uint acceptedTimestamp;
+        IMigration target;
+        address dataStore;
+    }
+
+    AddressResolver resolver = AddressResolver(0);
+
+    mapping(bytes32 => Migration) proposals;
+
+    constructor(address _owner) public Owned(_owner) {
+        waitingPeriod = 3 hours;
+    }
+
+    function setWaitingPeriod(uint _waitingPeriod) external onlyOwner {
+        waitingPeriod = _waitingPeriod;
+    }
+
+    function propose(
+        bytes32 version,
+        IMigration target,
+        address dataStore
+    )
+        external
+    {
+        // Anyone can call
+        require(proposals[version].target == IMigration(0), "Cannot modify existing proposal");
+        require(address(target) != address(0), "Invalid target");
+        proposals[version] = Migration(
+                0,
+                target,
+                dataStore
+            );
+    }
+
+    function accept(bytes32 version) external onlyOwner {
+        require(proposals[version].target != IMigration(0), "invalid proposal");
+        proposals[version].acceptedTimestamp = now;
+    }
+
+    function reject(bytes32 version) external onlyOwner {
+        require(proposals[version].target != IMigration(0), "invalid proposal");
+        delete proposals[version];
+    }
+
+    function execute(bytes32 version, uint index) external onlyOwner {
+        Migration memory migration = proposals[version];
+        require(migration.acceptedTimestamp > 0, "Must be accepted to execute");
+        require(now > migration.acceptedTimestamp + waitingPeriod, "Waiting period not yet expired");
```

Comment:
There could be an additional requirement here, which checks that the incoming index has not already been executed. In fact, it could even be a little stricter; it could check that the specified index makes sense. I.e. if step 0 was just executed, and index 2 is specified it would fail with a message like "Invalid migration step index. Step 1 needs to be executed before step 2".